File: clients/ios/JarvisClient/JarvisClient/ViewModels/VoiceViewModel.swift
// VoiceViewModel.swift
// Jarvis iOS Client — clients/ios/JarvisClient/JarvisClient/ViewModels/
//
// The single source of truth for all voice interaction state.
// Owns and coordinates:
//   AudioCaptureEngine  → microphone frames
//   WakeWordDetector    → "Hey Jarvis" trigger
//   GRPCVoiceService    → bidirectional backend stream
//
// State machine:
//   idle ──[wake word]──▶ listening ──[end of speech / VAD]──▶ processing
//     ▲                                                              │
//     └────────────── idle ◀── speaking ◀───────────────────────────┘
//
// SwiftUI views observe @Published properties directly.
// No business logic lives in views.

import Foundation
import SwiftUI
import AVFoundation
import os.log

// MARK: - Voice State

/// Mirrors StatusEvent.State from voice.proto — drives HUD rendering.
public enum VoiceState: Equatable {
    case idle
    case listening
    case processing
    case speaking
    case error(String)

    var displayLabel: String {
        switch self {
        case .idle:          return "Standby"
        case .listening:     return "Listening…"
        case .processing:    return "Processing…"
        case .speaking:      return "Speaking…"
        case .error(let m):  return "Error: \(m)"
        }
    }

    var isActive: Bool {
        self == .listening || self == .processing || self == .speaking
    }
}

// MARK: - HUD Action Model

/// Decoded client-side representation of a HUDAction proto message.
public struct HUDActionModel: Identifiable {
    public let id = UUID()
    public let type: HUDActionType
    public let payloadJSON: String
    public let severity: HUDSeverity
    public let receivedAt: Date
}

public enum HUDActionType: String {
    case openApp         = "TYPE_OPEN_APP"
    case showCard        = "TYPE_SHOW_CARD"
    case setTimer        = "TYPE_SET_TIMER"
    case navigate        = "TYPE_NAVIGATE"
    case dismissHUD      = "TYPE_DISMISS_HUD"
    case dispatchAgent   = "TYPE_DISPATCH_AGENT"
    case hardwareCmd     = "TYPE_HARDWARE_CMD"
    case securityProtocol = "TYPE_SECURITY_PROTOCOL"
    case unknown         = "TYPE_UNKNOWN"
}

public enum HUDSeverity {
    case info, warning, critical, emergency

    var color: Color {
        switch self {
        case .info:      return .blue
        case .warning:   return .yellow
        case .critical:  return .orange
        case .emergency: return .red
        }
    }
}

// MARK: - Transcript Model

public struct TranscriptLine: Identifiable {
    public let id = UUID()
    public let text: String
    public let isFinal: Bool
    public let confidence: Float
    public let timestamp: Date
    public let source: TranscriptSource
}

public enum TranscriptSource {
    case user       // STT from microphone
    case jarvis     // NLP reply text
}

// MARK: - VoiceViewModel

@MainActor
public final class VoiceViewModel: ObservableObject {

    // MARK: - Published: Core state

    /// Current voice pipeline state — primary driver of HUD rendering.
    @Published public private(set) var voiceState: VoiceState = .idle

    /// True when the gRPC stream is connected to the backend.
    @Published public private(set) var isConnected: Bool = false

    /// True when AudioCaptureEngine is running (mic active).
    @Published public private(set) var isMicActive: Bool = false

    // MARK: - Published: Audio

    /// Live RMS energy from the mic — drives the waveform view [0, 1].
    @Published public private(set) var micRMS: Float = 0.0

    // MARK: - Published: Transcript

    /// The current in-progress (possibly partial) user utterance.
    @Published public private(set) var liveTranscript: String = ""

    /// Full conversation history shown in the HUD transcript view.
    @Published public private(set) var transcriptHistory: [TranscriptLine] = []

    // MARK: - Published: Reply

    /// The most recent Jarvis reply text.
    @Published public private(set) var lastReply: String = ""

    /// True when Jarvis needs confirmation before acting.
    @Published public private(set) var requiresConfirmation: Bool = false

    /// The intent label from the last NLP response.
    @Published public private(set) var lastIntent: String = ""

    // MARK: - Published: HUD Actions

    /// Pending HUD actions the view layer should render or execute.
    @Published public private(set) var pendingActions: [HUDActionModel] = []

    // MARK: - Published: TTS Audio

    /// Accumulated TTS audio data for the current reply (plays when isFinalChunk).
    @Published public private(set) var ttsAudioData: Data? = nil

    // MARK: - Published: Errors

    /// Most recent non-fatal error message for the HUD error badge.
    @Published public private(set) var lastErrorMessage: String? = nil

    // MARK: - Dependencies

    private let grpcService: GRPCVoiceService
    private let audioEngine: AudioCaptureEngine
    private let wakeWordDetector: WakeWordDetector
    private let ttsPlayer: TTSAudioPlayer

    // MARK: - Identity

    private let userID: String
    private let sessionID: String

    // MARK: - Internal state

    /// True after wake word fires, before end-of-speech — gates audio streaming.
    private var isStreaming: Bool = false

    /// Accumulates TTS chunks until isFinalChunk.
    private var ttsBuffer: Data = Data()

    private let log = Logger(subsystem: "com.jarvis.client", category: "VoiceViewModel")

    // MARK: - Init

    public init(
        userID: String,
        sessionID: String = UUID().uuidString,
        grpcConfiguration: VoiceServiceConfiguration = .development
    ) {
        self.userID    = userID
        self.sessionID = sessionID

        self.ttsPlayer = TTSAudioPlayer()

        self.grpcService = GRPCVoiceService(
            configuration: grpcConfiguration,
            onEvent: { _ in }   // replaced below after init
        )

        self.audioEngine     = AudioCaptureEngine()
        self.wakeWordDetector = WakeWordDetector(configuration: .default)

        wireCallbacks()
    }
}

// MARK: - Public Interface

extension VoiceViewModel {

    /// Starts the full pipeline: mic → wake word → gRPC stream.
    public func start() async {
        log.info("VoiceViewModel starting — user: \(self.userID) session: \(self.sessionID)")

        do {
            // 1. Start wake word detector first (needs speech permission).
            try await wakeWordDetector.start()

            // 2. Start mic capture.
            try await audioEngine.start()
            isMicActive = true

            // 3. Connect gRPC stream (StreamConfig sent immediately).
            await grpcService.connect(userID: userID, sessionID: sessionID)

            voiceState = .idle
            log.info("Pipeline ready — waiting for wake word")

        } catch let err as AudioCaptureError {
            handleLocalError("Mic error: \(err.localizedDescription)")
        } catch let err as WakeWordDetectorError {
            handleLocalError("Wake word error: \(err.localizedDescription)")
        } catch {
            handleLocalError(error.localizedDescription)
        }
    }

    /// Stops the pipeline gracefully.
    public func stop() async {
        log.info("VoiceViewModel stopping")

        wakeWordDetector.stop()
        await audioEngine.stop()
        await grpcService.disconnect()

        isStreaming  = false
        isMicActive  = false
        isConnected  = false
        voiceState   = .idle
        micRMS       = 0
        liveTranscript = ""
    }

    /// Manually signals end-of-speech (e.g. push-to-talk button release).
    public func endSpeech() async {
        guard isStreaming else { return }
        isStreaming = false
        await grpcService.sendEndOfSpeech()
        log.debug("Manual end-of-speech sent")
    }

    /// Cancels the current in-flight pipeline.
    public func cancel() async {
        isStreaming = false
        voiceState = .idle
        liveTranscript = ""
        await grpcService.sendCancel()
    }

    /// Confirms a pending action when requiresConfirmation == true.
    public func confirm() async {
        requiresConfirmation = false
        await grpcService.sendNewTurn()
    }

    /// Dismisses the top pending HUD action.
    public func dismissTopAction() {
        guard !pendingActions.isEmpty else { return }
        pendingActions.removeFirst()
    }

    /// Clears transcript history.
    public func clearHistory() {
        transcriptHistory.removeAll()
    }
}

// MARK: - Private: Callback Wiring

private extension VoiceViewModel {

    func wireCallbacks() {
        wireAudioEngine()
        wireWakeWordDetector()
        wireGRPCService()
    }

    // MARK: AudioCaptureEngine

    func wireAudioEngine() {
        audioEngine.onFrame = { [weak self] frame in
            guard let self else { return }

            // Always update the waveform RMS.
            self.micRMS = frame.rmsEnergy

            // Feed every frame into the wake word detector.
            self.wakeWordDetector.feed(frame)

            // Only stream audio upstream after the wake word has fired.
            guard self.isStreaming else { return }

            await self.grpcService.sendAudioChunk(
                data:            frame.pcmData,
                capturedAt:      frame.capturedAt,
                isWakeWordFrame: false
            )
        }

        audioEngine.onError = { [weak self] error in
            self?.handleLocalError(error.localizedDescription)
        }
    }

    // MARK: WakeWordDetector

    func wireWakeWordDetector() {
        wakeWordDetector.onDetection = { [weak self] detection in
            guard let self else { return }
            guard self.voiceState == .idle else { return }  // ignore re-fires while active

            self.log.info("Wake word: \"\(detection.matchedPhrase)\" — opening stream")
            self.isStreaming   = true
            self.liveTranscript = ""
            self.voiceState    = .listening

            // Pause TTS playback if Jarvis is mid-sentence.
            self.ttsPlayer.stop()

            // Send the trigger frame marked as wake word — server skips VAD gate.
            await self.grpcService.sendAudioChunk(
                data:            detection.triggerFrame.pcmData,
                capturedAt:      detection.triggerFrame.capturedAt,
                isWakeWordFrame: true
            )
        }

        wakeWordDetector.onError = { [weak self] error in
            self?.handleLocalError(error.localizedDescription)
        }
    }

    // MARK: GRPCVoiceService

    func wireGRPCService() {
        // Re-create the service wiring now that self exists.
        // GRPCVoiceService stores the closure at init — we use a forwarding
        // pattern so VoiceViewModel is the single handler.
        grpcService.onEvent = { [weak self] event in
            guard let self else { return }
            self.handleVoiceEvent(event)
        }

        // Mirror GRPCVoiceService.isConnected into our own published property.
        grpcService.$isConnected
            .receive(on: RunLoop.main)
            .assign(to: &$isConnected)
    }
}

// MARK: - Private: VoiceEvent Handler

private extension VoiceViewModel {

    func handleVoiceEvent(_ event: VoiceEvent) {
        switch event {

        case .statusChanged(let state, let message):
            handleStatusChange(state, message: message)

        case .transcriptReceived(let text, let isFinal, let confidence):
            handleTranscript(text: text, isFinal: isFinal, confidence: confidence)

        case .replyReceived(let replyText, let intent, let requiresConfirmation):
            handleReply(text: replyText, intent: intent, requiresConfirmation: requiresConfirmation)

        case .audioReplyReceived(let data, let text, let isFinalChunk):
            handleAudioReply(data: data, text: text, isFinalChunk: isFinalChunk)

        case .hudActionReceived(let action):
            handleHUDAction(action)

        case .streamError(let error):
            handleLocalError(error.localizedDescription)

        case .streamClosed:
            voiceState  = .idle
            isStreaming = false
            log.info("Stream closed")
        }
    }

    // MARK: Status

    func handleStatusChange(_ state: JarvisStatusEvent.State, message: String) {
        log.debug("Status → \(String(describing: state))")
        switch state {
        case .idle:
            voiceState  = .idle
            isStreaming = false
            // Resume mic for next wake word.
            try? audioEngine.resume()

        case .listening:
            voiceState = .listening

        case .processing:
            voiceState  = .processing
            isStreaming = false   // stop sending audio — utterance is committed
            liveTranscript = ""  // clear partial; final arrives via transcript event

        case .speaking:
            voiceState = .speaking
            // Pause mic to avoid capturing TTS output.
            audioEngine.pause()

        case .error:
            voiceState  = .error(message.isEmpty ? "Unknown error" : message)
            isStreaming = false
            lastErrorMessage = message

        case .ended:
            voiceState  = .idle
            isStreaming = false

        default:
            break
        }
    }

    // MARK: Transcript

    func handleTranscript(text: String, isFinal: Bool, confidence: Float) {
        liveTranscript = isFinal ? "" : text

        if isFinal && !text.isEmpty {
            let line = TranscriptLine(
                text:       text,
                isFinal:    true,
                confidence: confidence,
                timestamp:  Date(),
                source:     .user
            )
            appendTranscript(line)
            log.info("Final transcript: \"\(text)\" (conf: \(String(format: "%.2f", confidence)))")
        }
    }

    // MARK: Reply

    func handleReply(text: String, intent: String, requiresConfirmation: Bool) {
        lastReply                 = text
        lastIntent                = intent
        self.requiresConfirmation = requiresConfirmation

        let line = TranscriptLine(
            text:       text,
            isFinal:    true,
            confidence: 1.0,
            timestamp:  Date(),
            source:     .jarvis
        )
        appendTranscript(line)
        log.info("Reply: \"\(text)\" intent=\(intent) confirm=\(requiresConfirmation)")
    }

    // MARK: Audio Reply (TTS)

    func handleAudioReply(data: Data, text: String, isFinalChunk: Bool) {
        ttsBuffer.append(data)

        if isFinalChunk {
            let audio = ttsBuffer
            ttsBuffer = Data()
            ttsAudioData = audio
            ttsPlayer.play(audio)
            log.debug("TTS playback started — \(audio.count) bytes")
        }
    }

    // MARK: HUD Action

    func handleHUDAction(_ proto: JarvisHUDAction) {
        let typeStr  = proto.type.description   // e.g. "TYPE_SHOW_CARD"
        let actionType = HUDActionType(rawValue: typeStr) ?? .unknown
        let severity   = mapSeverity(proto.severity)

        let model = HUDActionModel(
            type:        actionType,
            payloadJSON: proto.payloadJson,
            severity:    severity,
            receivedAt:  Date()
        )

        pendingActions.insert(model, at: 0)   // newest first

        // Auto-dismiss non-critical actions after 8 seconds.
        if severity != .critical && severity != .emergency {
            Task {
                try? await Task.sleep(nanoseconds: 8_000_000_000)
                if let idx = pendingActions.firstIndex(where: { $0.id == model.id }) {
                    pendingActions.remove(at: idx)
                }
            }
        }

        // Immediately execute dismissal actions.
        if actionType == .dismissHUD {
            pendingActions.removeAll()
        }

        log.info("HUD action: \(typeStr) severity=\(String(describing: severity))")
    }

    // MARK: Errors

    func handleLocalError(_ message: String) {
        log.error("VoiceViewModel error: \(message)")
        voiceState       = .error(message)
        lastErrorMessage = message
        isStreaming      = false
    }

    // MARK: Helpers

    func appendTranscript(_ line: TranscriptLine) {
        transcriptHistory.append(line)
        // Cap history to avoid unbounded memory growth.
        if transcriptHistory.count > 200 {
            transcriptHistory.removeFirst(transcriptHistory.count - 200)
        }
    }

    func mapSeverity(_ proto: JarvisCommonSeverity) -> HUDSeverity {
        switch proto {
        case .info:      return .info
        case .warning:   return .warning
        case .critical:  return .critical
        case .emergency: return .emergency
        default:         return .info
        }
    }
}

// MARK: - TTSAudioPlayer

/// Lightweight AVAudioPlayer wrapper for TTS AudioReply payloads.
/// Kept private to VoiceViewModel — only VoiceViewModel drives playback.
private final class TTSAudioPlayer: NSObject, AVAudioPlayerDelegate {

    private var player: AVAudioPlayer?
    private let log = Logger(subsystem: "com.jarvis.client", category: "TTSAudioPlayer")

    func play(_ data: Data) {
        stop()
        do {
            player = try AVAudioPlayer(data: data)
            player?.delegate = self
            player?.prepareToPlay()
            player?.play()
        } catch {
            log.error("TTSAudioPlayer failed: \(error.localizedDescription)")
        }
    }

    func stop() {
        player?.stop()
        player = nil
    }

    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        self.player = nil
    }
}

// MARK: - GRPCVoiceService onEvent setter

// GRPCVoiceService stores its callback at init. We expose a settable
// property here so VoiceViewModel can wire itself after both objects exist.
extension GRPCVoiceService {
    /// Settable event callback — used by VoiceViewModel to replace the
    /// no-op closure passed at init time.
    var onEvent: ((VoiceEvent) -> Void)? {
        get { _onEvent }
        set { _onEvent = newValue ?? { _ in } }
    }
}

File: clients/android/app/src/main/kotlin/com/jarvis/client/viewmodel/VoiceViewModel.kt
// VoiceViewModel.kt
// Jarvis Android Client — viewmodel/
//
// Single source of truth for all voice interaction state.
// Owns and coordinates:
//   AudioCaptureEngine → microphone frames
//   WakeWordDetector   → "Hey Jarvis" trigger
//   GRPCVoiceService   → bidirectional backend stream
//   TtsPlayer          → AudioTrack playback of TTS audio chunks
//
// State machine:
//   idle ──[wake word]──▶ listening ──[end of speech / VAD]──▶ processing
//     ▲                                                              │
//     └────────────── idle ◀── speaking ◀───────────────────────────┘
//
// Compose UI observes StateFlow / SharedFlow properties directly.

package com.jarvis.client.viewmodel

import android.app.Application
import android.content.pm.PackageManager
import androidx.core.content.ContextCompat
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.jarvis.client.audio.AudioCaptureEngine
import com.jarvis.client.audio.AudioFrame
import com.jarvis.client.grpc.*
import com.jarvis.client.wakeword.WakeWordDetector
import com.jarvis.client.wakeword.WakeWordConfiguration
import jarvis.voice.Voice.HUDAction
import jarvis.voice.Voice.StatusEvent
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import android.util.Log
import android.Manifest

private const val TAG = "VoiceViewModel"

// ── Voice State ───────────────────────────────────────────────────────────────

enum class VoiceState {
    IDLE, LISTENING, PROCESSING, SPEAKING, ERROR;

    val isActive: Boolean get() = this == LISTENING || this == PROCESSING || this == SPEAKING

    val displayLabel: String get() = when (this) {
        IDLE       -> "Standby"
        LISTENING  -> "Listening…"
        PROCESSING -> "Processing…"
        SPEAKING   -> "Speaking…"
        ERROR      -> "Error"
    }
}

// ── HUD Action Models ─────────────────────────────────────────────────────────

enum class HudActionType(val protoValue: Int) {
    UNKNOWN(0),
    OPEN_APP(1),
    SHOW_CARD(2),
    SET_TIMER(3),
    NAVIGATE(4),
    DISMISS_HUD(5),
    DISPATCH_AGENT(6),
    HARDWARE_CMD(7),
    SECURITY_PROTOCOL(8);

    companion object {
        fun from(value: Int) = entries.firstOrNull { it.protoValue == value } ?: UNKNOWN
    }
}

enum class HudSeverity { INFO, WARNING, CRITICAL, EMERGENCY }

data class HudActionModel(
    val id:          String = java.util.UUID.randomUUID().toString(),
    val type:        HudActionType,
    val payloadJson: String,
    val severity:    HudSeverity,
    val receivedAt:  Long = System.currentTimeMillis(),
)

// ── Transcript ────────────────────────────────────────────────────────────────

enum class TranscriptSource { USER, JARVIS }

data class TranscriptLine(
    val id:         String = java.util.UUID.randomUUID().toString(),
    val text:       String,
    val isFinal:    Boolean,
    val confidence: Float,
    val timestamp:  Long = System.currentTimeMillis(),
    val source:     TranscriptSource,
)

// ── VoiceViewModel ────────────────────────────────────────────────────────────

class VoiceViewModel(
    application: Application,
    private val grpcConfig:  VoiceServiceConfiguration = VoiceServiceConfiguration.Development,
    private val wakeConfig:  WakeWordConfiguration     = WakeWordConfiguration.Default,
) : AndroidViewModel(application) {

    // ── Published State ───────────────────────────────────────────────────────

    private val _voiceState = MutableStateFlow(VoiceState.IDLE)
    val voiceState: StateFlow<VoiceState> = _voiceState.asStateFlow()

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _isMicActive = MutableStateFlow(false)
    val isMicActive: StateFlow<Boolean> = _isMicActive.asStateFlow()

    private val _micRms = MutableStateFlow(0f)
    val micRms: StateFlow<Float> = _micRms.asStateFlow()

    private val _liveTranscript = MutableStateFlow("")
    val liveTranscript: StateFlow<String> = _liveTranscript.asStateFlow()

    private val _transcriptHistory = MutableStateFlow<List<TranscriptLine>>(emptyList())
    val transcriptHistory: StateFlow<List<TranscriptLine>> = _transcriptHistory.asStateFlow()

    private val _lastReply = MutableStateFlow("")
    val lastReply: StateFlow<String> = _lastReply.asStateFlow()

    private val _lastIntent = MutableStateFlow("")
    val lastIntent: StateFlow<String> = _lastIntent.asStateFlow()

    private val _requiresConfirmation = MutableStateFlow(false)
    val requiresConfirmation: StateFlow<Boolean> = _requiresConfirmation.asStateFlow()

    private val _pendingActions = MutableStateFlow<List<HudActionModel>>(emptyList())
    val pendingActions: StateFlow<List<HudActionModel>> = _pendingActions.asStateFlow()

    private val _lastErrorMessage = MutableStateFlow<String?>(null)
    val lastErrorMessage: StateFlow<String?> = _lastErrorMessage.asStateFlow()

    private val _hasMicPermission = MutableStateFlow(false)
    val hasMicPermission: StateFlow<Boolean> = _hasMicPermission.asStateFlow()

    // ── Internal State ────────────────────────────────────────────────────────

    private var isStreaming = false
    private var ttsBuffer   = ByteArray(0)

    // ── Dependencies ──────────────────────────────────────────────────────────

    private val grpcService     = GRPCVoiceService(grpcConfig, viewModelScope)
    private val audioEngine     = AudioCaptureEngine(viewModelScope)
    private val wakeWordDetector = WakeWordDetector(application, wakeConfig)
    private val ttsPlayer       = TtsPlayer()

    // User / session IDs — set before connecting.
    private var userId:    String = ""
    private var sessionId: String = java.util.UUID.randomUUID().toString()

    // ── Init ──────────────────────────────────────────────────────────────────

    init {
        checkMicPermission()
        wireFlows()
    }

    // ── Public Interface ──────────────────────────────────────────────────────

    /** Call with a user ID before start(). */
    fun setUser(id: String) { userId = id }

    /** Checks/updates RECORD_AUDIO permission state. */
    fun checkMicPermission() {
        val ctx = getApplication<Application>()
        _hasMicPermission.value = ContextCompat.checkSelfPermission(
            ctx, Manifest.permission.RECORD_AUDIO
        ) == PackageManager.PERMISSION_GRANTED
    }

    /** Starts the full pipeline: wake word → mic → gRPC stream. */
    fun start() {
        if (!_hasMicPermission.value) {
            handleError("Microphone permission not granted")
            return
        }
        viewModelScope.launch {
            try {
                wakeWordDetector.start()

                audioEngine.start()
                _isMicActive.value = true

                grpcService.connect(userId, sessionId)

                _voiceState.value = VoiceState.IDLE
                Log.i(TAG, "Pipeline ready — waiting for wake word")
            } catch (e: Exception) {
                handleError("Start failed: ${e.message}")
            }
        }
    }

    /** Stops the pipeline gracefully. */
    fun stop() {
        viewModelScope.launch {
            wakeWordDetector.stop()
            audioEngine.stop()
            grpcService.disconnect()

            isStreaming           = false
            _isMicActive.value    = false
            _isConnected.value    = false
            _voiceState.value     = VoiceState.IDLE
            _micRms.value         = 0f
            _liveTranscript.value = ""
            ttsPlayer.stop()
            Log.i(TAG, "Pipeline stopped")
        }
    }

    /** Manually signals end-of-speech (e.g. push-to-talk button release). */
    fun endSpeech() {
        if (!isStreaming) return
        viewModelScope.launch {
            isStreaming = false
            grpcService.sendEndOfSpeech()
        }
    }

    /** Cancels the current in-flight pipeline. */
    fun cancel() {
        isStreaming = false
        _voiceState.value = VoiceState.IDLE
        _liveTranscript.value = ""
        viewModelScope.launch { grpcService.sendCancel() }
    }

    /** Confirms a pending action when requiresConfirmation == true. */
    fun confirm() {
        _requiresConfirmation.value = false
        viewModelScope.launch { grpcService.sendNewTurn() }
    }

    /** Dismisses the most recent pending HUD action. */
    fun dismissTopAction() {
        val current = _pendingActions.value
        if (current.isNotEmpty()) {
            _pendingActions.value = current.drop(1)
        }
    }

    /** Clears transcript history. */
    fun clearHistory() { _transcriptHistory.value = emptyList() }

    // ── Private: Flow Wiring ──────────────────────────────────────────────────

    private fun wireFlows() {
        wireAudioEngine()
        wireWakeWordDetector()
        wireGrpcEvents()
        wireGrpcConnected()
    }

    private fun wireAudioEngine() {
        // Collect frames on Main — update RMS and forward to wake word + gRPC.
        audioEngine.frames
            .onEach { frame ->
                _micRms.value = frame.rmsEnergy
                wakeWordDetector.feed(frame)
                if (isStreaming) {
                    grpcService.sendAudioChunk(frame, isWakeWordFrame = false)
                }
            }
            .launchIn(viewModelScope)

        // RMS passthrough from engine's own state.
        audioEngine.currentRms
            .onEach { _micRms.value = it }
            .launchIn(viewModelScope)
    }

    private fun wireWakeWordDetector() {
        wakeWordDetector.detections
            .onEach { detection ->
                if (_voiceState.value != VoiceState.IDLE) return@onEach

                Log.i(TAG, "Wake word: \"${detection.matchedPhrase}\" — opening stream")
                isStreaming             = true
                _liveTranscript.value  = ""
                _voiceState.value      = VoiceState.LISTENING
                ttsPlayer.stop()

                // Send the trigger frame as a wake word frame — server skips VAD.
                detection.triggerFrame?.let { frame ->
                    grpcService.sendAudioChunk(frame, isWakeWordFrame = true)
                }
            }
            .launchIn(viewModelScope)

        wakeWordDetector.errors
            .onEach { Log.w(TAG, "WakeWordDetector error: $it") }
            .launchIn(viewModelScope)
    }

    private fun wireGrpcEvents() {
        grpcService.events
            .onEach { event -> handleVoiceEvent(event) }
            .launchIn(viewModelScope)
    }

    private fun wireGrpcConnected() {
        grpcService.isConnected
            .onEach { _isConnected.value = it }
            .launchIn(viewModelScope)
    }

    // ── Private: VoiceEvent Handler ───────────────────────────────────────────

    private fun handleVoiceEvent(event: VoiceEvent) {
        when (event) {
            is VoiceEvent.StatusChanged      -> handleStatusChange(event.state, event.message)
            is VoiceEvent.TranscriptReceived -> handleTranscript(event.text, event.isFinal, event.confidence)
            is VoiceEvent.ReplyReceived      -> handleReply(event.replyText, event.intent, event.requiresConfirmation)
            is VoiceEvent.AudioReplyReceived -> handleAudioReply(event.data, event.text, event.isFinalChunk)
            is VoiceEvent.HudActionReceived  -> handleHudAction(event.action)
            is VoiceEvent.StreamError        -> handleError(event.error.toString())
            is VoiceEvent.StreamClosed       -> {
                _voiceState.value = VoiceState.IDLE
                isStreaming = false
                Log.i(TAG, "Stream closed")
            }
        }
    }

    private fun handleStatusChange(state: StatusEvent.State, message: String) {
        Log.d(TAG, "Status → $state")
        when (state) {
            StatusEvent.State.STATE_IDLE -> {
                _voiceState.value = VoiceState.IDLE
                isStreaming = false
                audioEngine.resume()
            }
            StatusEvent.State.STATE_LISTENING -> {
                _voiceState.value = VoiceState.LISTENING
            }
            StatusEvent.State.STATE_PROCESSING -> {
                _voiceState.value     = VoiceState.PROCESSING
                isStreaming           = false
                _liveTranscript.value = ""
            }
            StatusEvent.State.STATE_SPEAKING -> {
                _voiceState.value = VoiceState.SPEAKING
                // Pause mic during TTS to avoid capturing our own audio.
                audioEngine.pause()
            }
            StatusEvent.State.STATE_ERROR -> {
                _voiceState.value    = VoiceState.ERROR
                isStreaming          = false
                _lastErrorMessage.value = message.ifBlank { "Unknown error" }
            }
            StatusEvent.State.STATE_ENDED -> {
                _voiceState.value = VoiceState.IDLE
                isStreaming = false
            }
            else -> Unit
        }
    }

    private fun handleTranscript(text: String, isFinal: Boolean, confidence: Float) {
        _liveTranscript.value = if (isFinal) "" else text

        if (isFinal && text.isNotBlank()) {
            appendTranscript(TranscriptLine(
                text       = text,
                isFinal    = true,
                confidence = confidence,
                source     = TranscriptSource.USER,
            ))
            Log.i(TAG, "Final transcript: \"$text\" (conf: ${"%.2f".format(confidence)})")
        }
    }

    private fun handleReply(text: String, intent: String, requiresConfirmation: Boolean) {
        _lastReply.value               = text
        _lastIntent.value              = intent
        _requiresConfirmation.value    = requiresConfirmation

        appendTranscript(TranscriptLine(
            text       = text,
            isFinal    = true,
            confidence = 1f,
            source     = TranscriptSource.JARVIS,
        ))
        Log.i(TAG, "Reply: \"$text\" intent=$intent confirm=$requiresConfirmation")
    }

    private fun handleAudioReply(data: ByteArray, text: String, isFinalChunk: Boolean) {
        ttsBuffer += data
        if (isFinalChunk) {
            val audio = ttsBuffer
            ttsBuffer = ByteArray(0)
            ttsPlayer.play(audio)
            Log.d(TAG, "TTS playback started — ${audio.size} bytes")
        }
    }

    private fun handleHudAction(proto: HUDAction) {
        val type     = HudActionType.from(proto.type.number)
        val severity = mapSeverity(proto.severity.number)

        val model = HudActionModel(
            type        = type,
            payloadJson = proto.payloadJson,
            severity    = severity,
        )

        _pendingActions.value = listOf(model) + _pendingActions.value

        // Auto-dismiss non-critical actions after 8 seconds.
        if (severity != HudSeverity.CRITICAL && severity != HudSeverity.EMERGENCY) {
            viewModelScope.launch {
                delay(8_000)
                _pendingActions.value = _pendingActions.value.filterNot { it.id == model.id }
            }
        }

        if (type == HudActionType.DISMISS_HUD) {
            _pendingActions.value = emptyList()
        }

        Log.i(TAG, "HUD action: $type severity=$severity")
    }

    private fun handleError(message: String) {
        Log.e(TAG, "VoiceViewModel error: $message")
        _voiceState.value       = VoiceState.ERROR
        _lastErrorMessage.value = message
        isStreaming             = false
    }

    // ── Private: Helpers ──────────────────────────────────────────────────────

    private fun appendTranscript(line: TranscriptLine) {
        val current = _transcriptHistory.value.toMutableList()
        current.add(line)
        if (current.size > 200) current.removeAt(0)
        _transcriptHistory.value = current
    }

    private fun mapSeverity(protoSeverity: Int): HudSeverity = when (protoSeverity) {
        1    -> HudSeverity.INFO
        2    -> HudSeverity.WARNING
        3    -> HudSeverity.CRITICAL
        4    -> HudSeverity.EMERGENCY
        else -> HudSeverity.INFO
    }

    override fun onCleared() {
        super.onCleared()
        stop()
        ttsPlayer.release()
    }
}

// ── TtsPlayer ─────────────────────────────────────────────────────────────────

/**
 * Lightweight AudioTrack wrapper for TTS AudioReply payloads.
 * PCM-16LE at 16 kHz mono — matches the server TTS output format.
 */
private class TtsPlayer {

    private var track: android.media.AudioTrack? = null

    fun play(pcmData: ByteArray) {
        stop()
        if (pcmData.isEmpty()) return

        val minBuf = android.media.AudioTrack.getMinBufferSize(
            16_000,
            android.media.AudioFormat.CHANNEL_OUT_MONO,
            android.media.AudioFormat.ENCODING_PCM_16BIT,
        )
        val bufSize = maxOf(minBuf, pcmData.size)

        track = android.media.AudioTrack(
            android.media.AudioManager.STREAM_MUSIC,
            16_000,
            android.media.AudioFormat.CHANNEL_OUT_MONO,
            android.media.AudioFormat.ENCODING_PCM_16BIT,
            bufSize,
            android.media.AudioTrack.MODE_STATIC,
        ).also { t ->
            t.write(pcmData, 0, pcmData.size)
            t.play()
        }
    }

    fun stop() {
        track?.let { t ->
            if (t.playState == android.media.AudioTrack.PLAYSTATE_PLAYING) t.stop()
            t.flush()
            t.release()
        }
        track = null
    }

    fun release() = stop()
}
